01_fib/shahriys/main.go

```diff
@@ -0,0 +1,51 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var outfib io.Writer = os.Stdout
+
+// Function to calculate nth Fibonacci or Negafibonacci number
+func nfib(n int) int64 {
+
+	if n >= 0 {
+
+		switch n {
+		case 0, 1:
+			return 1
+		default:
+			return nfib(n-1) + nfib(n-2)
+		}
+	} else {
+
+		switch n {
+		case -1:
+			return 1
+		case -2:
+			return -1
+		default:
+			return nfib(n+2) - nfib(n+1)
+		}
+	}
+}
+
+// Function to printh first n Fibonacci or Negafibonacci numbers
+func fib(n int) {
+	if n >= 0 {
+		for i := 0; i < n; i++ {
+			fmt.Fprintln(outfib, nfib(i))
```

Comment:
Looks like there's a performance hit here with the for loop into recursion. Potentially you could create a look up / map for Fn numbers that have already been calculated to reduce the number of iterations nfib would run, but that's at the cost of storing a map, so it's give and take. 👍

Reply:
Cost of creating/storing a map and assignment/lookup is `O(1)`. The current recursive implementation is `O(2^n)`, then on top of that we do the calculation for each value from `0` to `n`, meaning the overall complexity is `O(n2^n)`. The map would eliminate the extra factor of `2^n`, so definitely worth it.
See my other review comments though, since you don't need to do this to achieve `O(n)`.

Reply:
Recursion replaced with loop